src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelBindingContext.cs

```diff
@@ -289,12 +285,54 @@ public IValueProvider ValueProvider
         /// Gets or sets a predicate which will be evaluated for each property to determine if the property
         /// is eligible for model binding.
         /// </summary>
-        public Func<ModelBindingContext, string, bool> PropertyFilter { get; set; }
+        public Func<IModelBindingContext, string, bool> PropertyFilter { get { return _state.PropertyFilter; } set { _state.PropertyFilter = value; } }
 
         /// <summary>
         /// Gets or sets the <see cref="ValidationStateDictionary"/>. Used for tracking validation state to
         /// customize validation behavior for a model object.
         /// </summary>
-        public ValidationStateDictionary ValidationState { get; set; }
+        public ValidationStateDictionary ValidationState { get { return _state.ValidationState; } set { _state.ValidationState = value; } }
+
+        public ModelBindingResult? Result
+        {
+            get
+            {
+                return _state.Result;
+            }
+            set
+            {
+#if DEBUG
+                if (value.HasValue && value.Value == default(ModelBindingResult))
+                {
+                    throw new ArgumentException(nameof(ModelBindingResult));
+                }
+#endif
+
+                _state.Result = value;
+            }
+        }
+
+
+        private struct State
```

Comment:
Well, that is the entire point of the PR - currently each object (the entire model graph including dictionary and list entries) being model bound allocates an additional model binding context. So the assignment of all of the fields from one class to the next is already happening. The reason it's a struct in a Stack is that each of those child contexts can now be fields in a contiguous array w/out allocations.